atoma-state/src/state_manager.rs

```diff
@@ -325,47 +341,229 @@ impl AtomaState {
     ///
     /// # Arguments
     ///
-    /// * `model` - The model name for the task.
+    /// * `model` - The name of the model to search for (e.g., "gpt-4", "llama-2")
+    /// * `is_confidential` - Whether to only return nodes that support confidential computing
     ///
     /// # Returns
     ///
-    /// - `Result<Option<CheapestNode>>>`: A result containing either:
-    ///  - `Ok(Some<CheapestNode>)`: A `CheapestNode` object representing the node setting with the cheapest price.
-    ///  - `Ok(None)`: If no task is found for the given model.
-    ///  - `Err(AtomaStateManagerError)`: An error if the database query fails or if there's an issue parsing the results.
+    /// Returns a `Result` containing:
+    /// - `Ok(Some(CheapestNode))` - If a valid node subscription is found
+    /// - `Ok(None)` - If no valid node subscription exists for the model
+    /// - `Err(AtomaStateManagerError)` - If a database error occurs
     ///
-    /// # Errors
+    /// The returned `CheapestNode` contains:
+    /// - The task's small ID
+    /// - The price per compute unit
+    /// - The maximum number of compute units supported
     ///
-    /// This function will return an error if the database query fails.
+    /// # Security
+    ///
+    /// When `is_confidential` is true, this method:
+    /// - Only returns nodes that have valid public keys
+    /// - Only considers tasks with security level 2
+    /// - Requires nodes to be registered in the node_public_keys table
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// # use atoma_state::AtomaState;
+    /// # use sqlx::PgPool;
+    ///
+    /// async fn find_cheapest_node(state: &AtomaState) -> anyhow::Result<()> {
+    ///     // Find cheapest non-confidential node for GPT-4
+    ///     let regular_node = state.get_cheapest_node_for_model("gpt-4", false).await?;
+    ///     
+    ///     // Find cheapest confidential node for GPT-4
+    ///     let confidential_node = state.get_cheapest_node_for_model("gpt-4", true).await?;
+    ///     
+    ///     Ok(())
+    /// }
+    /// ```
     #[instrument(level = "trace", skip_all, fields(%model))]
-    pub async fn get_cheapest_node_for_model(&self, model: &str) -> Result<Option<CheapestNode>> {
-        let node_settings = sqlx::query(
-            "SELECT tasks.task_small_id, price_per_compute_unit, max_num_compute_units 
-            FROM (SELECT * 
-                  FROM tasks 
-                  WHERE is_deprecated=false
-                    AND model_name = $1) AS tasks 
-            JOIN (SELECT * 
-                  FROM node_subscriptions
-                  WHERE valid = true) AS node_subscriptions 
-            ON tasks.task_small_id=node_subscriptions.task_small_id 
+    pub async fn get_cheapest_node_for_model(
+        &self,
+        model: &str,
+        is_confidential: bool,
+    ) -> Result<Option<CheapestNode>> {
+        let mut query = String::from(
+            r#"
```

Comment:
add a `todo` for the future to double check the performance of this query, since it's triple join it can blow up.

Reply:
How come ?

Reply:
depends how many nodes there are and for how many we buy stacks
but you are right, if we do truncate the settlet stacks it should be fine